Resolves #270: add support for --time-style=relative

syphar opened the pull request:
This adds showing relative timestamps for #270 . 

For the first version this uses [`timeago`](https://docs.rs/timeago/) from the issue. 

I didn't find tests covering this part of the codebase, I'm happy to add them with a better idea on what should be covered and how. 

Also I'm happy to change the format if that's something we want (the LS++ format also looks neat). 

Different colouring for different ages I would add separately.

syphar commented:
How it looks like on my machine: 

<img width="676" alt="grafik" src="https://user-images.githubusercontent.com/540890/166107790-d227d64d-7ce5-48ac-acd1-edc3d490c8b8.png">

anthraxx commented on src/output/time.rs:
```diff
@@ -113,6 +120,21 @@ fn long_zoned(time: SystemTime, zone: &TimeZone) -> String {
+#[allow(trivial_numeric_casts)]
+fn relative(time: SystemTime) -> String {
+    format!(
+        "{:>13}", 
```
This seems like an arbitrary value. instead of relying what seems like a working column size, it would make more sense to not pad in here but just return the formatted relative time and afterwards collect the maximum length to pad to that afterwards.

syphar replied:
after a little more digging in the codebase I found the correct place to right-align columns and change this.

anthraxx commented:
> How it looks like on my machine:
> 
> <img alt="grafik" width="676" src="https://user-images.githubusercontent.com/540890/166107790-d227d64d-7ce5-48ac-acd1-edc3d490c8b8.png">

I'm wondering if it would be possible to omit the repetitive `" ago"`, this just consumes a lot of space without adding value to the output

syphar commented:
"ago" is gone now: 
<img width="728" alt="image" src="https://user-images.githubusercontent.com/540890/166109586-322ef8b7-c604-4175-ad6f-f6ffd6cbf6f4.png">

ariasuni commented:
> I didn't find tests covering this part of the codebase, I'm happy to add them with a better idea on what should be covered and how.

You can check the [Testing with vagrant](https://github.com/ogham/exa#testing-with-vagrant) section of the README, and [the README in the xtests folder](https://github.com/ogham/exa/tree/master/xtests). You can probably extend [these date tests](https://github.com/ogham/exa/blob/master/xtests/details-view-dates.toml) (tip: use e.g. `exa -l --time-style=relative --color=always > xtests/outputs/dates_long_timestyle_relative.ansitxt` to get the output to test).

ariasuni commented on src/options/view.rs:
```diff
@@ -257,6 +257,9 @@ impl TimeFormat {
         if &word == "default" {
             Ok(Self::DefaultFormat)
         }
+        else if &word == "relative" {
+            Ok(Self::Relative)
+        }
         else if &word == "iso" {
             Ok(Self::ISOFormat)
         }
```
Let’s use a match here.

syphar replied:
I updated the code, this needs a conversion to `str` to look nice-ish.

syphar commented:
> 



> > I didn't find tests covering this part of the codebase, I'm happy to add them with a better idea on what should be covered and how.
> 
> You can check the [Testing with vagrant](https://github.com/ogham/exa#testing-with-vagrant) section of the README, and [the README in the xtests folder](https://github.com/ogham/exa/tree/master/xtests). You can probably extend [these date tests](https://github.com/ogham/exa/blob/master/xtests/details-view-dates.toml) (tip: use e.g. `exa -l --time-style=relative --color=always > xtests/outputs/dates_long_timestyle_relative.ansitxt` to get the output to test).

Thanks for the hint! Somehow I missed that part. 

Will add tests

planet36 commented:
How are relative future dates printed?
`lsd` prints "in ____" for future dates, in contrast to "____ ago" for past dates.

syphar commented:
> > I didn't find tests covering this part of the codebase, I'm happy to add them with a better idea on what should be covered and how.
> 
> You can check the [Testing with vagrant](https://github.com/ogham/exa#testing-with-vagrant) section of the README, and [the README in the xtests folder](https://github.com/ogham/exa/tree/master/xtests). You can probably extend [these date tests](https://github.com/ogham/exa/blob/master/xtests/details-view-dates.toml) (tip: use e.g. `exa -l --time-style=relative --color=always > xtests/outputs/dates_long_timestyle_relative.ansitxt` to get the output to test).

@ariasuni I added tests as described.  Many other tests were failing on my machine, is that intended?

syphar commented:
> How are relative future dates printed?
> `lsd` prints "in ****" for future dates, in contrast to "**** ago" for past dates.

Currently it would print _just now_, also since `timeago` doesn't support future times yet. From the source, LS++ also only renders past dates correctly.

anthraxx commented:
> > How are relative future dates printed?
> > `lsd` prints "in ****" for future dates, in contrast to "**** ago" for past dates.
> 
> Currently it would print _just now_, also since `timeago` doesn't support future times yet. From the source, LS++ also only renders past dates correctly.

I would personally consider sind more an edge case that shouldn't normally happen. printing `just now` on future dates sounds reasonable enough

vi commented:
> For the first version this uses timeago from the issue.

Note that `timeago` was developed specifically for #270.
If something is amiss in `timeago`, more features can be added (especially if it does not require translators to update their translations).

vi commented:
> omit the repetitive " ago"

Note that `timeago` crate has special simplified function [`format_5chars`](https://docs.rs/timeago/latest/timeago/fn.format_5chars.html) for compressed time intervals representations of predictable width. This one is English-only and is not supposed to be translated though.

syphar commented:
> > omit the repetitive " ago"
> 
> Note that `timeago` crate has special simplified function [`format_5chars`](https://docs.rs/timeago/latest/timeago/fn.format_5chars.html) for compressed time intervals representations of predictable width. This one is English-only and is not supposed to be translated though.

This probably boils down to a design decision: 
1. change to 5chars 
2. add language to relative format (I see other date formats are localized, so probably we could use `timeago` languages too, where supported).

vi commented:
> change to 5chars

It may make sense to provide both options depending on listing options. Terse, but less user-friendly look may be poor default mode.

ariasuni commented:
> @ariasuni I added tests as described. Many other tests were failing on my machine, is that intended?

Uh yeah since they’re not on the CI, there are some that got broken and I need to merge the fixes and stuff… if your tests work then it’s fine. Also, it would be nice to have a test for future dates, see there: https://github.com/ogham/exa/blob/dceca33779e92f91589027650f9938d5360e16ba/xtests/details-view-dates.toml#L28-L36

syphar commented:
> > @ariasuni I added tests as described. Many other tests were failing on my machine, is that intended?
> 
> Uh yeah since they’re not on the CI, there are some that got broken and I need to merge the fixes and stuff… if your tests work then it’s fine. Also, it would be nice to have a test for future dates, see there:

done, added the far future test.

syphar commented:
@ariasuni is there anything else missing? I'm happy to change where necessary :)

Hyoihz commented:
I'm having a problem where if I run `exa` without any flags, it displays my files one per line as if I passed the `-1 or --oneline` flag. 

![video](https://user-images.githubusercontent.com/71059595/177167591-b0a18de9-f8ef-41ac-9525-caec8a9fb504.gif)

Hyoihz commented:
I'm having troubles displaying files properly with exa. 

`exa` when run as is (without any flags), displays my files one per line as if I'd passed the `-1` or `--oneline` flag.

![out](https://user-images.githubusercontent.com/71059595/177170199-e6cff769-bdce-447c-a51e-cb9c41442058.gif)

But when I use exa that's from my package manager, it works normally as how `ls` and `lsd` was displayed.

syphar commented:
just went through old open PRs and came back to this one :) 

@ariasuni I would love to see this in exa, I'm happy to add anything that's needed to get this finished up

syphar commented:
Hello @ariasuni I just remembered that I still had this PR open, and would love to see this feature in `exa` :) 
Is there anything that is missing that I can do to get this to land? 

To start, I just rebased on the current master :)

varon commented:
This would be great to have.

ariasuni commented:
Closing this, since exa is unmaintained (see https://github.com/ogham/exa/issues/1243), and this has been done in [the active fork eza](https://github.com/eza-community/eza). Thanks!
